File: BitManipulation/LeetCode/Easy/SortBySetBits/SortBySetBits.cpp
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
#include<bits/stdc++.h>
using namespace std;
vector<int> sortByBits(vector<int> &arr)
{
    auto lambda = [&](int &a, int &b)
    {
        int cA = __builtin_popcount(a); // Takes O(1) time
        int cB = __builtin_popcount(b);
        if (cA == cB)
            return a < b;
        return cA < cB;
    };
    sort(arr.begin(), arr.end(), lambda);
    return arr;
}
int main(void) {
    
    return 0;
}

File: BitManipulation/LeetCode/Easy/SortBySetBits/SortBySetBits.java
// https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/

import java.util.Arrays;

public class SortBySetBits {
    public int[] sortByBits(int[] arr) {
        Integer[] boxedArr = new Integer[arr.length];
        for (int i = 0; i < arr.length; i++) {
            boxedArr[i] = arr[i];
        }
        Arrays.sort(boxedArr, (a, b) -> {
            int cA = Integer.bitCount(a);
            int cB = Integer.bitCount(b);
            if(cA == cB) return a.compareTo(b);
            return Integer.compare(cA, cB);
        });
        for (int i = 0; i < arr.length; i++) {
            arr[i] = boxedArr[i];
        }

        return arr;
    }
}
